### PR TITLE
merge operator opts: log_softmax, take, where, log, split

### PR DESCRIPTION
**Pull request by TaoLv**
## Description ##
Merge optimizations from opt-1.3.1 branch, including: 

- log_softmax
- take
- where
- log
- split

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

**Comment by TaoLv** [comment]
@ciyongch Please review.

File: src/operator/nn/softmax-inl.h [hunk 1]
```diff
@@ -310,6 +313,54 @@ void SoftmaxCompute(const nnvm::NodeAttrs& attrs,
...
+  vsExp(n, in, dst);
+}
+
+static inline void log_softmax_parallel(TShape sh, int axis,
+                                        float * __restrict__ in, float * __restrict__ out) {
+  int64_t outer_size = prod(sh.data(), 0, axis);
+  int64_t channels = sh[axis];
+  // int inner_size = prod(n, axis+1, 4);
+
+#pragma omp parallel for
```
Comment (ciyongch) [inline src/operator/nn/softmax-inl.h]: Better to get recommended num_threads here :)

File: src/operator/nn/softmax-inl.h [hunk 2]
```diff
@@ -310,6 +313,54 @@ void SoftmaxCompute(const nnvm::NodeAttrs& attrs,
...
 
+#if MSHADOW_USE_MKL == 1
+static inline int64_t prod(int64_t* n, int start, int end) {
+  int64_t res = 1;
+  for (int i = start; i < end; i++)
+    res *= n[i];
+  return res;
+}
+
+static inline void max_(int64_t n, float * __restrict__ in, float *dst) {
```
Comment (ciyongch) [inline src/operator/nn/softmax-inl.h:324]: cblas_isamax() might have better perf.

**Comment by ciyongch** [comment]
Minor comment.
Looks like MKLLog is not picked into 1.4 release, waiting for MKL match patch?

**Comment by TaoLv** [comment]
Thank you :) MKLLog was not merged before 1.4 code freeze and we thought it's not critical enough to be picked.

File: src/operator/nn/softmax-inl.h (hunk 1 above)
Comment (TaoLv) [reply src/operator/nn/softmax-inl.h]: Sure. Fixed.

File: src/operator/nn/softmax-inl.h (hunk 2 above)
Comment (TaoLv) [reply src/operator/nn/softmax-inl.h:324]: Seems cblas_isamax() is for absolute values. Not sure how to use it here. Any suggestion?

https://software.intel.com/en-us/mkl-developer-reference-c-cblas-i-amax

**Comment by TaoLv** [comment]
I will merge this PR tomorrow if there is no other comment. @ciyongch @lihaofd @pengzhao-intel

File: src/operator/nn/softmax-inl.h (hunk 2 above)
Comment (ciyongch) [reply src/operator/nn/softmax-inl.h:324]: You're right, looks like no available max/min function for both signed and unsigned values from MKL. :)

**Review by ciyongch — APPROVED** [review]
